### PR TITLE
fix(CORS): CORS on git smart http protocol can not work. fixes #16350

### LINKED ISSUE #16350: CORS on git smart http protocol can not work on main branch
- Gitea version (or commit ref): 27c15782352aabaf9279c9c187d3658649d81e2f
- Git version: 2.25.1
- Operating system:
  ```
  Distributor ID:	Ubuntu
  Description:	Ubuntu 20.04.2 LTS
  Release:	20.04
  Codename:	focal
  ```
  
  
  
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
- Log gist:

## Description

This feature  #5719(#5700) can not work now.

The browser say:

```
Access to XMLHttpRequest at 'http://localhost:3000/name/repo.git/info/refs?service=git-upload-pack' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```
  
And the response header has nothing CORS header return.

The server report:

```
Completed OPTIONS /name/repo.git/info/refs?service=git-upload-pack 405 Method Not Allowed in 780.91µs
```

I've found it was incorrectly rejected  before entering httpBase.

https://github.com/go-gitea/gitea/blob/27c15782352aabaf9279c9c187d3658649d81e2f/routers/web/web.go#L150

The `context.Contexter()` reject it.

Just add the following code before `context.Contexter()` to fix:

```go
	common = append(common, CorsHandler) // added here
	// Removed: toolbox.Toolboxer middleware will provide debug informations which seems unnecessary
	common = append(common, context.Contexter())
```

### PR DESCRIPTION
**Pull request by snowyu**

### CODE REVIEW

File: routers/web/web.go [hunk 1]
```diff
@@ -1006,17 +1006,17 @@ func RegisterRoutes(m *web.Route) {
...
+				m.Get("/info/refs", CorsHandler(), repo.GetInfoRefs)
+				m.Get("/HEAD", CorsHandler(), repo.GetTextFile("HEAD"))
+				m.Get("/objects/info/alternates", CorsHandler(), repo.GetTextFile("objects/info/alternates"))
+				m.Get("/objects/info/http-alternates", CorsHandler(), repo.GetTextFile("objects/info/http-alternates"))
+				m.Get("/objects/info/packs", CorsHandler(), repo.GetInfoPacks)
+				m.Get("/objects/info/{file:[^/]*}", CorsHandler(), repo.GetTextFile(""))
+				m.Get("/objects/{head:[0-9a-f]{2}}/{hash:[0-9a-f]{38}}", CorsHandler(), repo.GetLooseObject)
+				m.Get("/objects/pack/pack-{file:[0-9a-f]{40}}.pack", CorsHandler(), repo.GetPackFile)
+				m.Get("/objects/pack/pack-{file:[0-9a-f]{40}}.idx", CorsHandler(), repo.GetIdxFile)
 			}, ignSignInAndCsrf)
```
Comment (lunny) [inline routers/web/web.go:1038]: Just add it after `ignSignInAdnCsrf` is better.

Comment (snowyu) [reply routers/web/web.go:1038]: I've tried this. But it does not work. MUST PUT `CorsHandler` in first.

Comment (zeripath) [reply routers/web/web.go:1038]: Then move the CorsHandler call first... 

```
m.Group("", CorsHandler(), func() { ... }, ignSignInAndCsrf) 
```

File: routers/web/web.go [hunk 2]
```diff
@@ -146,6 +147,23 @@ func Routes() *web.Route {
...
+			AllowedMethods: setting.CORSConfig.Methods,
+			AllowedHeaders: []string{"*"},
+			// OptionsPassthrough: true,
+			Debug:            true,
+			AllowCredentials: setting.CORSConfig.AllowCredentials,
+			MaxAge:           int(setting.CORSConfig.MaxAge.Seconds()),
+		})
+		common = append(common, corsHandle)
+	}
+	//*/
```
Comment (zeripath) [inline routers/web/web.go:165]: Why this section if you're adding it specifically below?

File: routers/web/web.go [hunk 3]
```diff
@@ -60,6 +60,7 @@ func CorsHandler() func(next http.Handler) http.Handler {
 			AllowedOrigins: setting.CORSConfig.AllowDomain,
 			//setting.CORSConfig.AllowSubdomain // FIXME: the cors middleware needs allowSubdomain option
 			AllowedMethods:   setting.CORSConfig.Methods,
+			AllowedHeaders:   []string{"*"},
```
Comment (zeripath) [inline routers/web/web.go:63]: Mostly we've just added the correct headers we need so in general I think this is a mistake.

**Comment by zeripath** [comment]
Please could you try #16496 instead.

**Pull request closed without merging**